sql: v22.2.1: error in argument for ×: unsupported OID 2206

This issue was autofiled by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry link: https://sentry.io/organizations/cockroach-labs/issues/3839460375/?referrer=webhooks_plugin

Panic message:

>encoding.go:853: error in argument for ×: unsupported OID 2206 with format code FormatBinary
(1) candidate pg code: 08P01
Wraps: (2) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind.func2
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:429
  | [...repeated from below...]
Wraps: (3) error in argument for ×
Wraps: (4) assertion failure
Wraps: (5) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase.DecodeDatum
  | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase/encoding.go:853
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind.func2
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:422
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:438
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2037
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1854
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
  | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
  | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728
  | runtime.goexit
  | 	src/runtime/asm_arm64.s:1172
Wraps: (6) unsupported OID 2206 with format code FormatBinary
Error types: (1) *pgerror.withCandidateCode (2) *withstack.withStack (3) *errutil.withPrefix (4) *assert.withAssertionFailure (5) *withstack.withStack (6) *errutil.leafError
-- report composition:
*errutil.leafError: unsupported OID 2206 with format code FormatBinary
encoding.go:853: *withstack.withStack (top exception)
*assert.withAssertionFailure
*errutil.withPrefix: error in argument for ×
conn_executor_prepare.go:429: *withstack.withStack (1)
*pgerror.withCandidateCode: 08P01
(check the extra data payloads)


<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

https://github.com/cockroachdb/cockroach/blob/cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92/pkg/sql/conn_executor_prepare.go#L428-L430 in pkg/sql.(*connExecutor).execBind.func2
https://github.com/cockroachdb/cockroach/blob/cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92/pkg/sql/conn_executor_prepare.go#L437-L439 in pkg/sql.(*connExecutor).execBind
https://github.com/cockroachdb/cockroach/blob/cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92/pkg/sql/conn_executor.go#L2036-L2038 in pkg/sql.(*connExecutor).execCmd
https://github.com/cockroachdb/cockroach/blob/cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92/pkg/sql/conn_executor.go#L1853-L1855 in pkg/sql.(*connExecutor).run
https://github.com/cockroachdb/cockroach/blob/cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92/pkg/sql/conn_executor.go#L831-L833 in pkg/sql.(*Server).ServeConn
https://github.com/cockroachdb/cockroach/blob/cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92/pkg/sql/pgwire/conn.go#L727-L729 in pkg/sql/pgwire.(*conn).processCommandsAsync.func1
src/runtime/asm_arm64.s#L1171-L1173 in runtime.goexit
</details>

```
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind.func2 at line 429
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind at line 438
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 2037
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 1854
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 832
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommandsAsync.func1 at line 728
src/runtime/asm_arm64.s in runtime.goexit at line 1172
```

| Tag | Value |
|---|---|
| Cockroach Release | `v22.2.1` |
| Cockroach SHA: | cf1e7e6bc6ef9e55510b9ed13bd068bb3894cd92 |
| Platform | linux arm64 |
| Distribution | CCL |
| Environment | v22.2.1 |
| Command | start-single-node |
| Go Version | ``|
| # of CPUs ||
| # of Goroutines ||


Jira issue: CRDB-22843

--- a/pkg/sql/logictest/testdata/logic_test/pg_catalog
+++ b/pkg/sql/logictest/testdata/logic_test/pg_catalog
@@ -1631,9 +1631,9 @@ oid     typname                typnamespace  typowner    typlen  typbyval  typty
 21      int2                   4294967127    NULL        2       true      b
 22      int2vector             4294967127    NULL        -1      false     b
 23      int4                   4294967127    NULL        4       true      b
-24      regproc                4294967127    NULL        8       true      b
+24      regproc                4294967127    NULL        4       true      b
 25      text                   4294967127    NULL        -1      false     b
-26      oid                    4294967127    NULL        8       true      b
+26      oid                    4294967127    NULL        4       true      b
 30      oidvector              4294967127    NULL        -1      false     b
 700     float4                 4294967127    NULL        4       true      b
 701     float8                 4294967127    NULL        8       true      b
@@ -1676,9 +1676,9 @@ oid     typname                typnamespace  typowner    typlen  typbyval  typty
 1562    varbit                 4294967127    NULL        -1      false     b
 1563    _varbit                4294967127    NULL        -1      false     b
 1700    numeric                4294967127    NULL        -1      false     b
-2202    regprocedure           4294967127    NULL        8       true      b
-2205    regclass               4294967127    NULL        8       true      b
-2206    regtype                4294967127    NULL        8       true      b
+2202    regprocedure           4294967127    NULL        4       true      b
+2205    regclass               4294967127    NULL        4       true      b
+2206    regtype                4294967127    NULL        4       true      b
 2207    _regprocedure          4294967127    NULL        -1      false     b
 2210    _regclass              4294967127    NULL        -1      false     b
 2211    _regtype               4294967127    NULL        -1      false     b
@@ -1695,9 +1695,9 @@ oid     typname                typnamespace  typowner    typlen  typbyval  typty
 3645    _tsquery               4294967127    NULL        -1      false     b
 3802    jsonb                  4294967127    NULL        -1      false     b
 3807    _jsonb                 4294967127    NULL        -1      false     b
-4089    regnamespace           4294967127    NULL        8       true      b
+4089    regnamespace           4294967127    NULL        4       true      b
 4090    _regnamespace          4294967127    NULL        -1      false     b
-4096    regrole                4294967127    NULL        8       true      b
+4096    regrole                4294967127    NULL        4       true      b
 4097    _regrole               4294967127    NULL        -1      false     b
 90000   geometry               4294967127    NULL        -1      false     b
 90001   _geometry              4294967127    NULL        -1      false     b

--- a/pkg/sql/pgwire/pgwirebase/encoding.go
+++ b/pkg/sql/pgwire/pgwirebase/encoding.go
@@ -556,12 +556,6 @@ func DecodeDatum(
 			}
 			i := int64(binary.BigEndian.Uint64(b))
 			return tree.NewDInt(tree.DInt(i)), nil
-		case oid.T_oid:
-			if len(b) < 4 {
-				return nil, pgerror.Newf(pgcode.Syntax, "oid requires 4 bytes for binary format")
-			}
-			u := binary.BigEndian.Uint32(b)
-			return tree.NewDOid(oid.Oid(u)), nil
 		case oid.T_float4:
 			if len(b) < 4 {
 				return nil, pgerror.Newf(pgcode.Syntax, "float4 requires 4 bytes for binary format")
@@ -832,6 +826,17 @@ func DecodeDatum(
 			}
 			if typ.Family() == types.TupleFamily {
 				return decodeBinaryTuple(ctx, evalCtx, b)
+			}
+			if typ.Family() == types.OidFamily {
+				if len(b) < 4 {
+					return nil, pgerror.Newf(pgcode.ProtocolViolation, "oid requires 4 bytes for binary format")
+				}
+				u := binary.BigEndian.Uint32(b)
+				oidTyp := types.Oid
+				if t, ok := types.OidToType[id]; ok {
+					oidTyp = t
+				}
+				return tree.NewDOidWithType(oid.Oid(u), oidTyp), nil
 			}
 		}
 	default:

--- a/pkg/sql/pgwire/testdata/pgtest/oid
+++ b/pkg/sql/pgwire/testdata/pgtest/oid
@@ -1,0 +1,35 @@
+# [2205, 4089, 24, 2206, 26] are the OIDs for regclass, regnamespace, regproc, regtype, and oid.
+send
+Parse {"Query": "SELECT $1, $2, $3, $4, $5", "ParameterOIDs": [2205,4089,24,2206,26]}
+Describe {"ObjectType": "S"}
+Bind {"ParameterFormatCodes": [1,1,1,1,1], "Parameters": [{"binary":"01000029"},{"binary":"0100002a"},{"binary":"0100002b"},{"binary":"0100002c"},{"binary":"ffffffff"}]}
+Execute
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"ParameterDescription","ParameterOIDs":[2205,4089,24,2206,26]}
+{"Type":"RowDescription","Fields":[{"Name":"?column?","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":2205,"DataTypeSize":4,"TypeModifier":-1,"Format":0},{"Name":"?column?","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":4089,"DataTypeSize":4,"TypeModifier":-1,"Format":0},{"Name":"?column?","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":24,"DataTypeSize":4,"TypeModifier":-1,"Format":0},{"Name":"?column?","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":2206,"DataTypeSize":4,"TypeModifier":-1,"Format":0},{"Name":"?column?","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":26,"DataTypeSize":4,"TypeModifier":-1,"Format":0}]}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"16777257"},{"text":"16777258"},{"text":"16777259"},{"text":"16777260"},{"text":"4294967295"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# [2205] is the OID for regclass.
+send
+Parse {"Query": "SELECT $1", "ParameterOIDs": [2205]}
+Bind {"ParameterFormatCodes": [1], "Parameters": [{"binary":"0029"}]}
+Execute
+Sync
+----
+
+until
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"ErrorResponse","Code":"08P01"}
+{"Type":"ReadyForQuery","TxStatus":"I"}

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -5766,7 +5766,7 @@ var baseDatumTypeSizes = map[types.Family]struct {
 	types.JsonFamily:           {unsafe.Sizeof(DJSON{}), variableSize},
 	types.UuidFamily:           {unsafe.Sizeof(DUuid{}), fixedSize},
 	types.INetFamily:           {unsafe.Sizeof(DIPAddr{}), fixedSize},
-	types.OidFamily:            {unsafe.Sizeof(DInt(0)), fixedSize},
+	types.OidFamily:            {unsafe.Sizeof(DOid{}.Oid), fixedSize},
 	types.EnumFamily:           {unsafe.Sizeof(DEnum{}), variableSize},
 
 	types.VoidFamily: {sz: unsafe.Sizeof(DVoid{}), variable: fixedSize},

